Add `sort_nodes.py`

mr0re1 opened the pull request:
````sh
$ /slurm/scripts/sort_nodes.py --help
usage: sort_nodes.py [-h] [--nodelist NODELIST] [--ntasks-per-node NTASKS_PER_NODE] [--output OUTPUT]

This script sorts nodes based on their `physicalHost`.

See https://cloud.google.com/compute/docs/instances/use-compact-placement-policies

You can reduce latency in tightly coupled HPC workloads (including distributed ML training)
by deploying them to machines that are located close together.
For example, if you deploy your workload on a single physical rack, you can expect lower latency
than if your workload is spread across multiple racks.
Sending data across multiple rack requires sending data through additional network switches.

Example usage:
``` my_sbatch.sh
#SBATCH --ntasks-per-node=8
#SBATCH --nodes=64

export SLURM_HOSTFILE=$(/slurm/scripts/sort_nodes.py)

srun -l hostname | sort
```

options:
  -h, --help            show this help message and exit
  --nodelist NODELIST   Slurm 'hostlist expression' of nodes to sort, if not set the value of SLURM_NODELIST environment variable will be used
  --ntasks-per-node NTASKS_PER_NODE
                        Number of times to repeat each node in resulting sorted list.
                        If not set, the value of SLURM_NTASKS_PER_NODE environment variable will be used,
                        if neither is set, defaults to 1
  --output OUTPUT       Output file to write, defaults to 'hosts.<uuid>'
````

samskillman commented on community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/slurm_files/scripts/tests/test_topology.py:
```diff
@@ -89,3 +91,20 @@ def tpu_se(ns: TstNodeset) -> TstTPU:
     conf.gen_topology_conf(util.Lookup(cfg))
     want_written = PRELUDE + "\n".join(want_compressed) + "\n\n"
     assert open(cfg.output_dir + "/cloud_topology.conf").read() == want_written
+
```
Suggest another test here demonstrating sorting of something like slurm-[10,5,12,7,22].

samskillman reviewed (changes requested):
Would suggest adding a test of numerical hostnames, and needs fix to get the lowest number first. Could be combined with a test of two groups of three nodes at depth 2, where the lowest number node is in the node with a higher first level string. something like 

x/slurm-1
x/slurm-2
x/slurm-3
y/slurm-0
y/slurm-4
z/slurm-10

I'd expect 0,4,10,1,2,3

mr0re1 commented:
@samskillman  , for 
> something like
x/slurm-1
x/slurm-2
x/slurm-3
y/slurm-0
y/slurm-4
z/slurm-10
I'd expect 0,4,10,1,2,3

The order `0, 4, 1, 2, 3, 10` seems to be more natural is the particular reason you wrote `0,4,10,1,2,3`?

mr0re1 commented:
* Addressed comments about name-order:
Don't use naive lexicographic order anywhere, always rely on order of names in `nodelist`.
Significantly refactor `def order`.

* Rename `topologicaly_sort_nodes -> sort_nodes` to reduce confusion, since order is not "topological".

samskillman commented:
> @samskillman , for
> 
> > something like
> > x/slurm-1
> > x/slurm-2
> > x/slurm-3
> > y/slurm-0
> > y/slurm-4
> > z/slurm-10
> > I'd expect 0,4,10,1,2,3
> 
> The order `0, 4, 1, 2, 3, 10` seems to be more natural is the particular reason you wrote `0,4,10,1,2,3`?

I agree - I meant for that z to be a y. I think either would be acceptable solutions in principle, but I agree 0,4,1,2,3,10 is better. My mistake on the typo.

nick-stroud commented on community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/slurm_files/scripts/sort_nodes.py:
```diff
@@ -0,0 +1,183 @@
+    if len(parts) >= 4:
+        return [*parts, node_name]
+    elif len(parts) == 3:
+        return [zone, *parts, node_name]  # add zone
```
It is not intuitive to me what is the significance of being >=4 or ==3. Do some physical hosts include zone? I am cautious about the special casing here since I expect physical host may change. At min can you add comment as to what you are checking for here, as you have done for other cases above.

mr0re1 replied:
Made it more robust by removing assertion about length of `parts`.

nick-stroud commented on community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/slurm_files/scripts/sort_nodes.py:
```diff
@@ -0,0 +1,183 @@
+    The resulting order will always start from the first node in the input list.
+    The ordering is "stable" with respect to the input order of the leafs i.e.
+    given a choice between two nodes (identical in other ways) it will select "nodelist-smallest" one.
+    """
```
From description and signature, the interface is not clear to me. It took some digging to understand that it is returning a list of node names (leaves) and that the node names are expected to be the last element in a path, I had been thinking strictly physical_host.

Something like the following would be helpful:
```
return: a list containing each leaf of the input paths, ordered as described above.  
```

nick-stroud commented on community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/slurm_files/scripts/sort_nodes.py:
```diff
@@ -0,0 +1,183 @@
+
+    # propagate "lexicographical key" from the leaves to the root
+    def set_lex_key(v: Vert) -> None:
+        if not v.children: # this is a node
```
nit: s/node/leaf/

nick-stroud commented on community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/slurm_files/scripts/sort_nodes.py:
```diff
@@ -0,0 +1,183 @@
+
+def order(paths: List[List[str]]) -> List[str]:
+    """
+    Orders the leafs of the tree in a way that minimizes the sum of distance in between 
```
s/leafs/leaves/

nick-stroud commented on community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/slurm_files/scripts/sort_nodes.py:
```diff
@@ -0,0 +1,183 @@
+            for u in v.children.values():
+                set_lex_key(u)
+            v.lex_key = min(u.lex_key for u in v.children.values())
+    set_lex_key(root)
```
I don't see where lex_key is ever used, which I think is what you want as dict maintains arg input order where desired.

mr0re1 replied:
Good catch. Looking into it, lead to refactoring, see the diff.
